src/GPUifyLoops.jl

```diff
@@ -20,54 +20,98 @@ export @setup, @loop, @synchronize
 export @scratch, @shmem
 export contextualize
 export @unroll
+export @launch
 
-@init @require CUDAnative="be33ccc6-a3ff-5ff2-a52e-74243cff1e17" begin
-    using .CUDAnative
-end
+##
+# contextualize
+##
 include("context.jl")
 
-iscpu(::GPU) = false
-iscpu(::CPU) = true
-sync(::CPU) = nothing
-sync(::CUDA) = sync_threads()
+macro launch(ex...)
+    # destructure the `@launch` expression
+    call = ex[end]
+    kwargs = ex[2:end-1]
+
+    device = ex[1]
+
+    # destructure the kernel call
+    if call.head != :call
+        throw(ArgumentError("second argument to @launch should be a function call"))
+    end
+
+    f = call.args[1]
+    args = call.args[2:end]
+
+    quote
+        $launch($esc(device), $esc(f), $(map(esc, args)...); $(map(esc, kwargs)...))
+    end
+end
 
-@deprecate iscpu(::Val{:GPU}) iscpu(CUDA())
-@deprecate iscpu(::Val{:CPU}) iscpu(CPU())
-@deprecate sync(::Val{:GPU}) sync(CUDA())
-@deprecate sync(::Val{:CPU}) sync(CPU())
 
 
 """
-    @setup Dev
+   launch(::Device, f, args..., kwargs...)
 
-Setups some hidden state within the function that allows the other macros to
-properly work.
+Launch a kernel on the GPU. `kwargs` are passed to `@cuda`
+`kwargs` can be any of the compilation and runtime arguments
+normally passed to `@cuda`.
+"""
+launch(::CPU, f, args...; kwargs...) = f(args...)
 
-```julia
-function kernel(::Dev, A) where {Dev}
-    @setup Dev
-    # ...
-end
+"""
+    launch_config(::F, maxthreads, args...; kwargs...)
 
-kernel(A::Array) = kernel(CPU(), A)
-kernel(A::CuArray) = @cuda kernel(CUDA(), A)
-```
+Calculate a valid launch configuration based on the typeof(F), the
+maximum number of threads, the functions arguments and the particular
+launch configuration passed to the call.
+
+Return a NamedTuple that has `blocks`, `threads`, `shmem`, and `stream`.
+All arguments are optional, but blocks and threads is recommended.
 """
-macro setup(sym)
-    esc(:(local __DEVICE = $sym()))
+function launch_config(@nospecialize(f), maxthreads, args...; kwargs...)
+    return kwargs
+end
+
+@init @require CUDAnative="be33ccc6-a3ff-5ff2-a52e-74243cff1e17" begin
+    using .CUDAnative
+
+    function launch(::CUDA, f::F, args...; kwargs...) where F
+        compiler_kwargs, call_kwargs = CUDAnative.split_kwargs(kwargs)
+        GC.@preserve args begin
+            kernel_args = map(cudaconvert, args)
+            kernel_tt = Tuple{map(Core.Typeof, kernel_args)...}
+            kernel = cufunction(contextualize(f), kernel_tt; compilation_kwargs)
+
+            maxthreads = CUDAnative.maxthreads(kernel)
+            config = launch_config(f, maxthreads, args...; call_kwargs)
+
+            kernel(kernel_args...; config...)
+        end
+        return nothing
+    end
 end
 
+isdevice() = false
+Cassette.overdub(ctx::Ctx, ::typeof(isdevice)) = true
```

Comment:
Forgive me for my ignorance, but what happens here? This function is changed to output true if the code is compiled for GPU, somehow?

Reply:
Yes basically. We use Cassette for contextual execution and within a contexts it is going to replace the original function with this one